### DISCUSSION TITLE
Custom hadoop rpc plugins

### DISCUSSION THREAD

**Discussion by yuyanlei-8130** [apache/skywalking#11394] [Q&A]
To monitor the links between hadoop clients and hadoop servers, I refer to https://skyapm.github.io/document-cn-translation-of-skywalking/zh/6.3.0/guides/Java-Plugin-Development-Guide.html,It seems that hadoop can't implement such a plugin, and the hadoop rpc is issued like this:
```
public void sendRpcRequest(final Call call) throws InterruptedException, IOException {
      if (shouldCloseConnection.get()) {
        return;
      }

      RpcRequestHeaderProto header = ProtoUtil.makeRpcRequestHeader(
          call.rpcKind, OperationProto.RPC_FINAL_PACKET, call.id, call.retry,
          clientId, call.alignmentContext);

      final ResponseBuffer buf = new ResponseBuffer();
      header.writeDelimitedTo(buf);
      RpcWritable.wrap(call.rpcRequest).writeTo(buf);

      synchronized (sendRpcRequestLock) {
        Future<?> senderFuture = sendParamsExecutor.submit(new Runnable() {
          @Override
          public void run() {
            try {
              synchronized (ipcStreams.out) {
                if (shouldCloseConnection.get()) {
                  return;
                }
                if (LOG.isDebugEnabled()) {
                  LOG.debug(getName() + " sending #" + call.id
                      + " " + call.rpcRequest);
                }
                // RpcRequestHeader + RpcRequest
                ipcStreams.sendRequest(buf.toByteArray());
                ipcStreams.flush();
              }
            } catch (IOException e) {
              // exception at this point would leave the connection in an
              // unrecoverable state (eg half a call left on the wire).
              // So, close the connection, killing any outstanding calls
              markClosed(e);
            } finally {
              //the buffer is just an in-memory buffer, but it is still polite to
              // close early
              IOUtils.closeStream(buf);
            }
          }
        });
      
        try {
          senderFuture.get();
        } catch (ExecutionException e) {
          Throwable cause = e.getCause();
          
          // cause should only be a RuntimeException as the Runnable above
          // catches IOException
          if (cause instanceof RuntimeException) {
            throw (RuntimeException) cause;
          } else {
            throw new RuntimeException("unexpected checked exception", cause);
          }
        }
      }
    }
```
proto is an rpc request sent asynchronously. In this case, how can I develop a plugin to monitor the rpc between hadoop client and hadoop server?

**Comment by wu-sheng** [comment]
If it is only a Hadoop client, I don't think you need to add headers. This plugin should work like a MySQL client, right?

**Reply by yuyanlei-8130** [reply]
For example, the hbase server will interact with the hadoop server through rpc. hbase is the client of hadoop. If I want to monitor the interaction between the hbase server and the hadoop server, I have to implement a hadoop plug-in. Monitor RPCS between hadoop clients and hadoop servers

**Reply by wu-sheng** [reply]
You didn't follow my question. Are you going to build a server plugin for Hadoop? If not, there is no requirement for manipulating the header. You could check MySQL plugin, there is no header relative code.

**Reply by yuyanlei-8130** [reply]
Yes, I'm going to build a server plugin for Hadoop

**Reply by yuyanlei-8130** [reply]
Is there a hadoop plugin? I couldn't find one

**Reply by wu-sheng** [reply]
No, and that would be not easy work to trace executions inside the Hadoop cluster. 
Why would you like to do this? 

In your question context, if an RPC doesn't have a header/metadata to add custom things, you have to change the body of the request. 
This would make your remote peer sensitive to the new protocol. This is inevitable

**Reply by yuyanlei-8130** [reply]
I wanted a quick fix for locating hadoop
Like I mentioned with hadoop rpc, I have to modify the hadoop source code to adapt to it?

**Reply by wu-sheng** [reply]
You need to be clear about your question, `what do you mean can't`, which part do you mean `can't`?

**Reply by wu-sheng** [reply]
async RPC call is not an issue. There are many aync RPC framework supported. And `proto` is only about serialization, not RPC.
So, I can't follow up the whole thing. I thought you mean header can't be injected.
